feat: improved indexer

jerome89 opened the pull request:
- Indexer now supports deferred write
  * Delayed write to spread search & write request for given amount of time.
- Indexer now preloads keys into the next index
  * To prevent heavy search & write request when index rolls to the next one.
- Use _id for sha-512 of metric key
  * This enables users to store metric keys greater 512-byte.
- Improved SeriesByTag function
  * Checks every argument is properly given to prevent unintended queries.
- Improved multiGetRequestContainer
  * Now it does not store from, until to create range indices.
  * Instead, it stores exact indices to create.
- Changed createIndexIfNotExists in ElasticsearchClient.kt
  * Takes Set of indices as an argument to create exact indices.
  * This prevents to create unintended indices given by from, until.

jerome89 committed:
feat: improved indexer

- Indexer now supports deferred write
  * Delayed write to spread search & write request for given amount of time.
- Indexer now preloads keys into the next index
  * To prevent heavy search & write request when index rolls to the next one.
- Use _id for sha-512 of metric key
  * This enables users to store metric keys greater 512-byte.
- Improved SeriesByTag function
  * Checks every argument is properly given to prevent unintended queries.
- Improved multiGetRequestContainer
  * Now it does not store from, until to create range indices.
  * Instead, it stores exact indices to create.
- Changed createIndexIfNotExists in ElasticsearchClient.kt
  * Takes Set of indices as an argument to create exact indices.
  * This prevents to create unintended indices given by from, until.

jerome89 committed:
feat: improved indexer

- fixed crashed test

jerome89 committed:
feat: improved indexer

- fixed crashed test(InfluxDbMetricConverterTest)

Dark0096 committed:
refactor: add test code after simpler modification of existing logic

Dark0096 committed:
refactor: add test code after simpler modification of existing logic

Dark0096 committed:
refactor: rename expression matcher method

Dark0096 commented:
I refactored the SeriesByTagFunction and added some test code! Please review. :)
https://github.com/graphene-monitoring/graphene/pull/81

Dark0096 commented on graphene-common/src/main/kotlin/com/graphene/common/utils/HashUtils.kt:
```diff
@@ -0,0 +1,43 @@
+ * @author Sam Clarke <www.samclarke.com>
+ * @license MIT
+ */
+object HashUtils {
```
👍

Dark0096 commented on graphene-writer/src/main/kotlin/com/graphene/writer/store/key/elasticsearch/ElasticsearchClient.kt:
```diff
@@ -1,15 +1,15 @@
 interface ElasticsearchClient : Closeable, RotatedIndexAware {
 
-  fun createIndexIfNotExists(index: String, tenant: String, from: Long?, to: Long?)
+  fun createIndexIfNotExists(indices: Set<String>)
```
It became a very simple interface. :)

jerome89 committed:
feat: improved indexer

- Upgraded dependencies.
  * Metrics that are exposed by micrometer having metricattribute tag.
- Removed unused codes.
- Added spring environment to base metric tags.
- Changed behavior of KeyStoreHandler.
  * Scheduler waits until the current job is finished
- Changed some properties in KeyStoreHandler.
  * Removed unused properties.
  * Use flushInterval as scheduler delay.

jerome89 committed:
feat: improved indexer

- Updated reader's properties

jerome89 committed:
Merge pull request #81 from graphene-monitoring/feature/improved-indexer_add-test

refactor: add test code after simpler modification of existing logic

Dark0096 commented on graphene-reader/src/main/kotlin/com/graphene/reader/store/key/elasticsearch/handler/TagBasedKeySearchHandler.kt:
```diff
@@ -40,11 +40,16 @@ class TagBasedKeySearchHandler(
-            for (tag in tags) {
-              path.addTag(tag.key, tag.value.toString())
+            for (source in sources) {
+              if (! INTERNAL_KEYS.contains(source.key)) {
```
It might be a stupid question. :)
When can this happen?

jerome89 replied:
It always happens because indexed keys are in the form: 
```
{
 "@key": "cpu_idle;host=a", 
 "@name":"cpu_idle",
 "@tags": ["host"], 
 "host":"a"
}
```
It is okay to put those values into Path() but excluded because I think it is useless.

Dark0096 replied:
I simply modified the logic for that and added the test code:
https://github.com/graphene-monitoring/graphene/pull/82/commits/2ddd93e29f497010f94dcd340ca427be7cda0bcc

Dark0096 commented on graphene-writer/src/main/kotlin/com/graphene/writer/store/key/elasticsearch/handler/AbstractElasticsearchKeyStoreHandler.kt:
```diff
@@ -30,124 +29,188 @@ abstract class AbstractElasticsearchKeyStoreHandler(
+      tenant,
+      grapheneMetric.timestampMillis() + (this.cacheExpireIntervalInSeconds * 1_000L)
+    )
+    if (originalIndex != preloadIndex && isProcessable(grapheneMetric) && keyCache.putIfAbsent("${preloadIndex}_${grapheneMetric.key}")) {
```
How about caching by id like above?

```suggestion
    if (originalIndex != preloadIndex && isProcessable(grapheneMetric) && keyCache.putIfAbsent("${preloadIndex}_${grapheneMetric.id}")) {
```

jerome89 replied:
Oh, that's important thing it was my mistake. Thanks for the advice.

jerome89 committed:
feat: improved indexer

- Fixed key caching bug when indexing

Dark0096 commented on graphene-writer/src/main/resources/application-local.yml:
```diff
@@ -118,9 +121,12 @@ graphene:
               index: "index-based-key-path"
               type: "path"
+              cacheExpireIntervalInSeconds: 300
+              targetProcessTime: 30000
               bulk:
                 actions: 10000
-                interval: 500
+                preloadActions: 5000
+                interval: 10000
```
How about renaming those settings to make them more readable? The term `PreLoad` simply feels like loading a specific key to memory in advance. Rather, it would be better to change the `PreIndexing` or another name that suits the actual operation.
